In [1]:
import pandas as pd

In [2]:
event_log = pd.read_csv('C:/Users/rrine/Desktop/Example MIX TYPES - Lana Labs.txt', delimiter="\t")

In [3]:
event_log.head()

TransID         ActivityName   Time eLetter_Type  eLetter_ID  ShapeID  \
0        1              Request    0.0       Type_1           1      767   
1        1              Request  600.0       Type_1           1      767   
2        1                Start  600.0       Type_1           1     1376   
3        1                Start  600.0       Type_1           1     1376   
4        1  Schedule Activities  600.0       Type_1           1       15   

   ShapeNumber  OK_After_Review_14_perc  OK_After_Review_17_perc  \
0            1                      100                      100   
1            1                      100                      100   
2            2                      100                      100   
3            2                      100                      100   
4            3                      100                      100   

   OK_After_Review_25_perc  ... Paste_In_Template_STD  Preempt  Priority  \
0                      100  ...                    10    False         0   
1                      100  ...                    10    False         0   
2                      100  ...                    10    False         0   
3                      100  ...                    10    False         0   
4                      100  ...                    10    False         0   

   Pull_Target_List_MAX  Pull_Target_List_MIN  Review_13_AVG  Review_13_STD  \
0                     2                     1             30             15   
1                     2                     1             30             15   
2                     2                     1             30             15   
3                     2                     1             30             15   
4                     2                     1             30             15   

   Review_16_AVG  Review_16_STD  Send_On  
0             30             15       10  
1             30             15       10  
2             30             15       10  
3             30             15       10  
4             30             15       10  

[5 rows x 37 columns]

In [4]:
len(event_log)

3267

We have 3267 events in total, as each row represents an event.

In [5]:
pd.Series(event_log['ActivityName']).value_counts()
# df2 = event_log.groupby("ActivityName").count()
# print(df2)

OK?                                       352
Review                                    247
Outline Text, Targeting, and URLs         190
Start                                     176
Schedule Activities                       176
Request                                   176
Broadcast Request for Related Articles    164
Send Email                                161
Delay                                     152
Gather Materials                          140
Internal Review And Changes               137
Optimize Graphics                         136
Paste In Template                         134
 Create Initial Draft                     125
Create Initial Target List                118
Create Proof                              117
Send On                                   114
Send To Production                        110
Analyze Criteria                          108
Pull Target List                          106
End                                       102
Modify Criteria                   

In [6]:
case_id = event_log['eLetter_ID']
len(case_id.unique())

88

We have 88 different cases.

In [7]:
event_log['eLetter_Type'].unique()

array(['Type_1', 'Type_2', 'Type_4'], dtype=object)

In [8]:
eletter_type = event_log['eLetter_Type']
pd.Series(eletter_type).value_counts()

Type_1    1598
Type_2     901
Type_4     768
Name: eLetter_Type, dtype: int64

The values of eLetter_Type attributes are 'object' and Type_1 occurs 1598 times often, Type_2 occurs 901 times often and Type_4 occurs 768 times often.

In [9]:
import datetime
fix_date_time = datetime.datetime(2016, 1, 1, 10, 0, 0)
print(fix_date_time)

2016-01-01 10:00:00


In [10]:
new_date_time = []
for i in event_log['Time']:
    updated_date_time = fix_date_time + datetime.timedelta(0,i)
    new_date_time.append(updated_date_time)
print(new_date_time)

[datetime.datetime(2016, 1, 1, 10, 0), datetime.datetime(2016, 1, 1, 10, 10), datetime.datetime(2016, 1, 1, 10, 10), datetime.datetime(2016, 1, 1, 10, 10), datetime.datetime(2016, 1, 1, 10, 10), datetime.datetime(2016, 1, 1, 10, 34, 22, 294853), datetime.datetime(2016, 1, 1, 10, 34, 22, 294853), datetime.datetime(2016, 1, 1, 10, 49, 22, 294853), datetime.datetime(2016, 1, 1, 10, 49, 22, 294853), datetime.datetime(2016, 1, 1, 12, 0), datetime.datetime(2016, 1, 1, 12, 10), datetime.datetime(2016, 1, 1, 12, 10), datetime.datetime(2016, 1, 1, 12, 10), datetime.datetime(2016, 1, 1, 12, 10), datetime.datetime(2016, 1, 1, 12, 26, 22, 455046), datetime.datetime(2016, 1, 1, 12, 26, 22, 455046), datetime.datetime(2016, 1, 1, 12, 41, 22, 455046), datetime.datetime(2016, 1, 1, 12, 41, 22, 455046), datetime.datetime(2016, 1, 1, 16, 0), datetime.datetime(2016, 1, 1, 16, 10), datetime.datetime(2016, 1, 1, 16, 10), datetime.datetime(2016, 1, 1, 16, 10), datetime.datetime(2016, 1, 1, 16, 10), datetime.

calculated timestamps for each row

In [11]:
event_log['TotalTime'] = new_date_time

In [12]:
event_log.sort_values(by=['ActivityName', 'eLetter_ID', 'TotalTime'], inplace=True)

In [13]:
event_log.head()

TransID           ActivityName           Time eLetter_Type  eLetter_ID  \
48        1   Create Initial Draft   86663.936862       Type_1           1   
84        1   Create Initial Draft  106382.944338       Type_1           1   
68        2   Create Initial Draft   88500.000000       Type_2           2   
86        2   Create Initial Draft  106890.731234       Type_2           2   
72        3   Create Initial Draft   90900.000000       Type_4           4   

    ShapeID  ShapeNumber  OK_After_Review_14_perc  OK_After_Review_17_perc  \
48      207            9                      100                      100   
84      207            9                      100                      100   
68      207            9                       80                       85   
86      207            9                       80                       85   
72      207            9                      100                      100   

    OK_After_Review_25_perc  ... Preempt  Priority  Pull_Target_List_MAX  \
48                      100  ...   False         0                     2   
84                      100  ...   False         0                     2   
68                       25  ...   False         0                     2   
86                       25  ...   False         0                     2   
72                      100  ...   False         0                     6   

    Pull_Target_List_MIN  Review_13_AVG  Review_13_STD  Review_16_AVG  \
48                     1             30             15             30   
84                     1             30             15             30   
68                     1             30             15             30   
86                     1             30             15             30   
72                     3             90             45             90   

    Review_16_STD  Send_On                  TotalTime  
48             15       10 2016-01-02 10:04:23.936862  
84             15       10 2016-01-02 15:33:02.944338  
68             15       10 2016-01-02 10:35:00.000000  
86             15       10 2016-01-02 15:41:30.731234  
72             45       30 2016-01-02 11:15:00.000000  

[5 rows x 38 columns]

Combining two subsequent rows for the same activity into one with start and complete timestamp

In [14]:
event_number = 0
new_columns =  ['EventNumber', 'ActivityName', 'eLetter_ID', 'eLetter_Type', 'StartTime', 'EndTime']
new_event_log = pd.DataFrame(columns = new_columns)
activityname = None
eletterid = None

event_log.iloc[[0]]
for index, row in event_log.iterrows():
    if index == 0:
        activityname = row['ActivityName']
        eletterid = row['eLetter_ID']
        
        new_event_log = new_event_log.append({
            'EventNumber': event_number,
            'ActivityName' : row['ActivityName'], 
            'eLetter_ID' : row['eLetter_ID'],
            'eLetter_Type': row['eLetter_Type'],
            'StartTime': row['TotalTime']
        } , ignore_index=True)
        event_number = event_number + 1
    else:
        if (row['ActivityName']==activityname and row['eLetter_ID']==eletterid):
            new_event_log.loc[new_event_log.index[-1], 'EndTime']= row['TotalTime']
        else:
            activityname = row['ActivityName']
            eletterid = row['eLetter_ID']
        
            event_number = event_number + 1
            new_event_log = new_event_log.append({
                'EventNumber': event_number,
                'ActivityName' : row['ActivityName'], 
                'eLetter_ID' : row['eLetter_ID'],
                'eLetter_Type': row['eLetter_Type'],
                'StartTime': row['TotalTime']
            } , ignore_index=True)

In [15]:
new_event_log

EventNumber           ActivityName eLetter_ID eLetter_Type  \
0              1   Create Initial Draft          1       Type_1   
1              2   Create Initial Draft          2       Type_2   
2              3   Create Initial Draft          4       Type_4   
3              4   Create Initial Draft          5       Type_1   
4              5   Create Initial Draft         13       Type_1   
...          ...                    ...        ...          ...   
1391        1392                  Start        341       Type_1   
1392        1393                  Start        349       Type_1   
1393        1394                  Start        350       Type_2   
1394        1395                  Start        352       Type_4   
1395        1396                  Start        353       Type_1   

                      StartTime                     EndTime  
0    2016-01-02 10:04:23.936862  2016-01-02 15:33:02.944338  
1    2016-01-02 10:35:00.000000  2016-01-02 15:41:30.731234  
2    2016-01-02 11:15:00.000000  2016-01-03 13:45:31.909560  
3    2016-01-02 10:25:00.000000  2016-01-02 15:05:20.828435  
4    2016-01-03 10:25:00.000000  2016-01-03 13:50:16.543405  
...                         ...                         ...  
1391 2016-01-29 18:10:00.000000         2016-01-29 18:10:00  
1392 2016-01-30 10:10:00.000000         2016-01-30 10:10:00  
1393 2016-01-30 12:10:00.000000         2016-01-30 12:10:00  
1394 2016-01-30 16:10:00.000000         2016-01-30 16:10:00  
1395 2016-01-30 18:10:00.000000         2016-01-30 18:10:00  

[1396 rows x 6 columns]

In [16]:
len(new_event_log)

1396

Removing all rows for which we do not have a matching second row

In [17]:
new_event_log.dropna(subset = ["EndTime"], inplace=True)

In [18]:
len(new_event_log)

1352

We have 1352 events.

In [19]:
pd.Series(new_event_log['ActivityName']).value_counts()

Schedule Activities                       88
Request                                   88
Start                                     88
Delay                                     76
Broadcast Request for Related Articles    76
Gather Materials                          64
Outline Text, Targeting, and URLs         63
 Create Initial Draft                     62
Internal Review And Changes               61
Optimize Graphics                         61
OK?                                       59
Review                                    59
Paste In Template                         59
Create Proof                              58
Send On                                   57
Send To Production                        55
Create Initial Target List                55
Send Email                                53
Pull Target List                          53
Analyze Criteria                          53
End                                       51
Modify Criteria                           11
Change    

This output shows the number of events for each ActivityName.

In [20]:
case_id_new = new_event_log['eLetter_ID']
len(case_id_new.unique())

88

In [21]:
new_event_log['eLetter_Type'].unique()

array(['Type_1', 'Type_2', 'Type_4'], dtype=object)

In [22]:
eletter_type_new = new_event_log['eLetter_Type']
pd.Series(eletter_type_new).value_counts()

Type_1    675
Type_2    352
Type_4    325
Name: eLetter_Type, dtype: int64

The values of eLetter_Type attributes are 'object' and Type_1 occurs 1598 times often, Type_2 occurs 901 times often and Type_4 occurs 768 times often.

Calculating the difference between start and complete time for each event and show the distribution

In [23]:
difference = new_event_log['StartTime'] - new_event_log['EndTime']
time_diff = difference.astype('timedelta64[h]')
new_event_log['Time_diff'] = time_diff

C:\Users\rrine\anaconda3\lib\site-packages\pandas\core\arrays\datetimelike.py:1345: PerformanceWarning: Adding/subtracting object-dtype array to DatetimeArray not vectorized
  PerformanceWarning,


In [24]:
new_event_log.head()

EventNumber           ActivityName eLetter_ID eLetter_Type  \
0           1   Create Initial Draft          1       Type_1   
1           2   Create Initial Draft          2       Type_2   
2           3   Create Initial Draft          4       Type_4   
3           4   Create Initial Draft          5       Type_1   
4           5   Create Initial Draft         13       Type_1   

                   StartTime                     EndTime  Time_diff  
0 2016-01-02 10:04:23.936862  2016-01-02 15:33:02.944338       -6.0  
1 2016-01-02 10:35:00.000000  2016-01-02 15:41:30.731234       -6.0  
2 2016-01-02 11:15:00.000000  2016-01-03 13:45:31.909560      -27.0  
3 2016-01-02 10:25:00.000000  2016-01-02 15:05:20.828435       -5.0  
4 2016-01-03 10:25:00.000000  2016-01-03 13:50:16.543405       -4.0

surprising element: timespent for each ActivityName

In [25]:
activity_timespent = new_event_log.groupby('ActivityName').agg({'Time_diff': ['sum']}).reset_index()

In [26]:
activity_timespent

ActivityName Time_diff
                                                 sum
0                     Create Initial Draft   -1593.0
1                         Analyze Criteria    -763.0
2   Broadcast Request for Related Articles   -3793.0
3                                   Change     -51.0
4               Create Initial Target List   -3076.0
5                             Create Proof    -363.0
6                                    Delay    -571.0
7                                      End       0.0
8                         Gather Materials   -3685.0
9              Internal Review And Changes   -1148.0
10                         Modify Criteria     -96.0
11                                     OK?   -2273.0
12                       Optimize Graphics   -1011.0
13       Outline Text, Targeting, and URLs     -80.0
14                       Paste In Template    -883.0
15                        Pull Target List    -452.0
16                                 Request     -88.0
17                                  Review   -1003.0
18                     Schedule Activities    -152.0
19                              Send Email   -3348.0
20                                 Send On    -165.0
21                      Send To Production     -55.0
22                                   Start       0.0

bonus question: Using twitter search api

In [27]:
fix_date_time = datetime.datetime(2020, 11, 1, 10, 0, 0) 
new_date_time = []
for i in event_log['Time']:
    updated_date_time = fix_date_time + datetime.timedelta(0,i)
    new_date_time.append(updated_date_time)
print(new_date_time)

[datetime.datetime(2020, 11, 2, 10, 4, 23, 936862), datetime.datetime(2020, 11, 2, 15, 33, 2, 944338), datetime.datetime(2020, 11, 2, 10, 35), datetime.datetime(2020, 11, 2, 15, 41, 30, 731234), datetime.datetime(2020, 11, 2, 11, 15), datetime.datetime(2020, 11, 3, 13, 45, 31, 909560), datetime.datetime(2020, 11, 2, 10, 25), datetime.datetime(2020, 11, 2, 15, 5, 20, 828435), datetime.datetime(2020, 11, 3, 10, 25), datetime.datetime(2020, 11, 3, 13, 50, 16, 543405), datetime.datetime(2020, 11, 3, 10, 35), datetime.datetime(2020, 11, 3, 16, 9, 4, 398857), datetime.datetime(2020, 11, 3, 11, 20, 34, 634009), datetime.datetime(2020, 11, 4, 15, 56, 14, 879615), datetime.datetime(2020, 11, 3, 17, 7, 23, 657716), datetime.datetime(2020, 11, 4, 11, 58, 55, 425826), datetime.datetime(2020, 11, 4, 17, 8, 4, 332226), datetime.datetime(2020, 11, 5, 12, 13, 21, 465206), datetime.datetime(2020, 11, 5, 10, 35), datetime.datetime(2020, 11, 5, 13, 56, 26, 205356), datetime.datetime(2020, 11, 5, 11, 13, 

In [28]:
event_log['new_date_time'] = new_date_time

In [29]:
event_log.sort_values(by=['ActivityName', 'eLetter_ID', 'new_date_time'], inplace=True)

In [30]:
event_number = 0
new_columns =  ['EventNumber', 'ActivityName', 'eLetter_ID', 'eLetter_Type', 'StartTime', 'EndTime']
new_event_log = pd.DataFrame(columns = new_columns)
activityname = None
eletterid = None

event_log.iloc[[0]]
for index, row in event_log.iterrows():
    if index == 0:
        activityname = row['ActivityName']
        eletterid = row['eLetter_ID']
        
        new_event_log = new_event_log.append({
            'EventNumber': event_number,
            'ActivityName' : row['ActivityName'], 
            'eLetter_ID' : row['eLetter_ID'],
            'eLetter_Type': row['eLetter_Type'],
            'StartTime': row['new_date_time']
        } , ignore_index=True)
        event_number = event_number + 1
    else:
        if (row['ActivityName']==activityname and row['eLetter_ID']==eletterid):
            new_event_log.loc[new_event_log.index[-1], 'EndTime']= row['new_date_time']
        else:
            activityname = row['ActivityName']
            eletterid = row['eLetter_ID']
        
            event_number = event_number + 1
            new_event_log = new_event_log.append({
                'EventNumber': event_number,
                'ActivityName' : row['ActivityName'], 
                'eLetter_ID' : row['eLetter_ID'],
                'eLetter_Type': row['eLetter_Type'],
                'StartTime': row['new_date_time']
            } , ignore_index=True)

In [31]:
new_event_log.dropna(subset = ["EndTime"], inplace=True)

In [32]:
new_event_log.head()

EventNumber           ActivityName eLetter_ID eLetter_Type  \
0           1   Create Initial Draft          1       Type_1   
1           2   Create Initial Draft          2       Type_2   
2           3   Create Initial Draft          4       Type_4   
3           4   Create Initial Draft          5       Type_1   
4           5   Create Initial Draft         13       Type_1   

                   StartTime                     EndTime  
0 2020-11-02 10:04:23.936862  2020-11-02 15:33:02.944338  
1 2020-11-02 10:35:00.000000  2020-11-02 15:41:30.731234  
2 2020-11-02 11:15:00.000000  2020-11-03 13:45:31.909560  
3 2020-11-02 10:25:00.000000  2020-11-02 15:05:20.828435  
4 2020-11-03 10:25:00.000000  2020-11-03 13:50:16.543405

In [33]:
from twython import Twython  
import json

with open("C:/Users/rrine/Desktop/twitter_credentials.json", "r") as file:  
    creds = json.load(file)

# Instantiate an object
python_tweets = Twython(creds['CONSUMER_KEY'], creds['CONSUMER_SECRET'])


# Create our query
query = {'q': 'penguins',  
        'result_type': 'popular',
        'count': 100,
        'lang': 'en',
        'until': '2020-11-29'
        }


# Search tweets
dict_ = {'user': [], 'date': [], 'text': [], 'favorite_count': []}  
for status in python_tweets.search(**query)['statuses']:  
    dict_['user'].append(status['user']['screen_name'])
    dict_['date'].append(status['created_at'])
    dict_['text'].append(status['text'])
    dict_['favorite_count'].append(status['favorite_count'])

# Structure data in a pandas DataFrame for easier manipulation
df = pd.DataFrame(dict_)  
df.sort_values(by='favorite_count', inplace=True, ascending=False) 
df.head(5)

import pytz
df['changed_date'] = df.apply(lambda row: datetime.datetime.strptime(row['date'],'%a %b %d %H:%M:%S +0000 %Y').replace(tzinfo=pytz.UTC), axis=1)
df['latest_date'] = df.apply(lambda row: row['changed_date'].strftime('%Y-%m-%d'), axis=1)
twitter = df.groupby('latest_date').agg({'favorite_count': ['sum']}).reset_index()
twitter.columns = ['latest_date', 'favorite_count']

In [34]:
twitter

latest_date  favorite_count
0  2020-11-23            3382
1  2020-11-24            1847
2  2020-11-27            4491
3  2020-11-28            7405

In [35]:
penguin_count = new_event_log.copy()
penguin_count['latest_date'] = penguin_count.apply(lambda row: row['StartTime'].strftime('%Y-%m-%d'), axis=1)

In [36]:
final_event_log = pd.merge(penguin_count, twitter, how='left', left_on=['latest_date'], right_on = ['latest_date'])

In [37]:
final_event_log.tail(20)

EventNumber ActivityName eLetter_ID eLetter_Type  \
1332        1377        Start        277       Type_1   
1333        1378        Start        278       Type_2   
1334        1379        Start        280       Type_4   
1335        1380        Start        281       Type_1   
1336        1381        Start        289       Type_1   
1337        1382        Start        290       Type_2   
1338        1383        Start        292       Type_4   
1339        1384        Start        293       Type_1   
1340        1385        Start        301       Type_1   
1341        1386        Start        302       Type_2   
1342        1387        Start        304       Type_4   
1343        1388        Start        305       Type_1   
1344        1389        Start        337       Type_1   
1345        1390        Start        338       Type_2   
1346        1391        Start        340       Type_4   
1347        1392        Start        341       Type_1   
1348        1393        Start        349       Type_1   
1349        1394        Start        350       Type_2   
1350        1395        Start        352       Type_4   
1351        1396        Start        353       Type_1   

                      StartTime                     EndTime latest_date  \
1332 2020-11-24 10:10:00.000000         2020-11-24 10:10:00  2020-11-24   
1333 2020-11-24 12:10:00.000000         2020-11-24 12:10:00  2020-11-24   
1334 2020-11-24 16:17:58.546407  2020-11-24 16:17:58.546407  2020-11-24   
1335 2020-11-24 18:31:25.515380  2020-11-24 18:31:25.515380  2020-11-24   
1336 2020-11-25 10:10:00.000000         2020-11-25 10:10:00  2020-11-25   
1337 2020-11-25 12:10:00.000000         2020-11-25 12:10:00  2020-11-25   
1338 2020-11-25 16:10:00.000000         2020-11-25 16:10:00  2020-11-25   
1339 2020-11-25 18:10:00.000000         2020-11-25 18:10:00  2020-11-25   
1340 2020-11-26 10:10:00.000000         2020-11-26 10:10:00  2020-11-26   
1341 2020-11-26 12:10:00.000000         2020-11-26 12:10:00  2020-11-26   
1342 2020-11-26 16:10:00.000000         2020-11-26 16:10:00  2020-11-26   
1343 2020-11-26 18:14:57.508286  2020-11-26 18:14:57.508286  2020-11-26   
1344 2020-11-29 10:10:00.000000         2020-11-29 10:10:00  2020-11-29   
1345 2020-11-29 12:10:00.000000         2020-11-29 12:10:00  2020-11-29   
1346 2020-11-29 16:10:00.000000         2020-11-29 16:10:00  2020-11-29   
1347 2020-11-29 18:10:00.000000         2020-11-29 18:10:00  2020-11-29   
1348 2020-11-30 10:10:00.000000         2020-11-30 10:10:00  2020-11-30   
1349 2020-11-30 12:10:00.000000         2020-11-30 12:10:00  2020-11-30   
1350 2020-11-30 16:10:00.000000         2020-11-30 16:10:00  2020-11-30   
1351 2020-11-30 18:10:00.000000         2020-11-30 18:10:00  2020-11-30   

      favorite_count  
1332          1847.0  
1333          1847.0  
1334          1847.0  
1335          1847.0  
1336             NaN  
1337             NaN  
1338             NaN  
1339             NaN  
1340             NaN  
1341             NaN  
1342             NaN  
1343             NaN  
1344             NaN  
1345             NaN  
1346             NaN  
1347             NaN  
1348             NaN  
1349             NaN  
1350             NaN  
1351             NaN